pandas/core/ops.py

```diff
@@ -136,6 +137,49 @@ def maybe_upcast_for_op(obj):
     return obj
 
 
+def unpack_and_defer(func):
+    """
+    Decorate a binary method to return NotImplemented where appropriate,
+    unpack arguments, and validate argument shapes.
+
+    Parameters
+    ----------
+    func : binary method
+
+    Returns
+    -------
+    decorated : binary method
+    """
+
+    @wraps(func)
+    def method(self, other):
+        other = lib.item_from_zerodim(other)
+        # TODO: maybe_upcast_for_op?
+
+        for abc_cls in [ABCDataFrame, ABCSeries, ABCIndexClass]:
+            if isinstance(self, abc_cls):
+                break
+            elif isinstance(other, abc_cls):
+                return NotImplemented
```

Comment:
I am with @jbrockmendel here, the decorator is ok,  though *maybe* have an extended name with this to avoid confusion.

``maybe_return_not_implemented`` or some such

Reply:
> To me, that makes the flow control a bit clearer. The downside is that every place you've added this decorator will need to include a

Yah I considered this, decided against it because of the verbosity.  If there turns out to be a consensus that this is preferred, I'll be OK with that.

Reply:
I won't object to the way you've written it here.